scene/2d/parallax_2d.cpp

```diff
@@ -0,0 +1,277 @@
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "parallax_2d.h"
+
+void Parallax2D::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_INTERNAL_PROCESS: {
+			if ((!repeat_size.x && !repeat_size.y) || (!autoscroll.x && !autoscroll.y)) {
+				return;
+			}
+
+			autoscroll_offset += autoscroll * get_process_delta_time();
+
+			real_t mod = Math::fmod(autoscroll_offset.x, repeat_size.x);
+			autoscroll_offset.x = mod < 0 ? mod + repeat_size.x : mod;
+			mod = Math::fmod(autoscroll_offset.y, repeat_size.y);
+			autoscroll_offset.y = mod < 0 ? mod + repeat_size.y : mod;
+
+			_update_scroll();
+		} break;
+
+		case NOTIFICATION_ENTER_TREE: {
+			group_name = "__cameras_" + itos(get_viewport_rid().get_id());
+			add_to_group(group_name);
+			_update_repeat();
+		} break;
+
+		case NOTIFICATION_EXIT_TREE: {
+			remove_from_group(group_name);
+		} break;
+	}
+}
+
+void Parallax2D::_camera_moved(const Transform2D &p_transform, const Point2 &p_screen_offset, const Point2 &p_adj_screen_pos) {
+	if (ignore_camera_scroll) {
+		return;
+	}
+
+	set_screen_offset(p_adj_screen_pos);
+}
+
+void Parallax2D::_update_scroll() {
+	if (!is_inside_tree() || Engine::get_singleton()->is_editor_hint()) {
+		return;
+	}
+
+	Point2 scroll_ofs = screen_offset;
+	Size2 vps = get_viewport_rect().size;
+
+	if (limit_begin.x <= limit_end.x - vps.x) {
+		if (screen_offset.x < limit_begin.x) {
+			scroll_ofs.x = limit_begin.x;
+		} else if (screen_offset.x + vps.x > limit_end.x) {
+			scroll_ofs.x = limit_end.x - vps.x;
+		}
```

Comment:
`screen_offset.x = CLAMP(limit_begin.x, limit_end.x - vps.x)`

Same for the y-component. Or just do both at the same time with Vector2.clamp()

Reply:
I should be able to use `CLAMP()`, but I wouldn't be able to use the `Vector2` method since each should only be clamped if the limits are the size of the screen or larger.

Reply:
Vector2's `clamp` limits each axis individually

Reply:
I know, I mean that there's another if check surrounding these lines referenced. It's possible for the set limit size for the x axis to be smaller than the screen but the y axis to be larger, meaning that the x shouldn't be clamped and the y should be.

Reply:
Then it actually does make sense as is, yeah...

Reply:
We can still clamp them separately, it's just that 
```cpp
if (limit_begin.x <= limit_end.x - vps.x) {
    scroll_ofs.x = CLAMP(scroll_ofs.x, limit_begin.x, limit_end.x - vps.x);
}

if (limit_begin.y <= limit_end.y - vps.y) {
    scroll_ofs.y = CLAMP(scroll_ofs.y, limit_begin.y, limit_end.y - vps.y);
}
```
has a different result than

```cpp
if (limit_begin <= limit_end - vps) {
    scroll_ofs = scroll_ofs.clamp(limit_begin, limit_end - vps);
}
```